server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java

```diff
@@ -66,7 +66,8 @@ private Long getEstimatedShardSize(ShardRouting shard) {
             return shardSizes.get(ClusterInfo.shardIdentifierFromRouting(shard.shardId(), true));
         } else {
             // initializing new (empty?) primary
-            return shard.getExpectedShardSize();
+            var priorSize = shardSizes.get(ClusterInfo.shardIdentifierFromRouting(shard));
+            return priorSize != null ? Math.max(priorSize, shard.getExpectedShardSize()) : shard.getExpectedShardSize();
```

Comment:
I think this branch could happen when the shard was unassigned for a short period of time (for example when node with only primary unexpectedly left the cluster and rejoined). In this case we should consider prior known shard size (I will open a separate pr for this).